File: android/src/main/java/com/navigine/naviginesdk/LocationViewController.java
package com.navigine.naviginesdk;

import android.content.Context;
import android.graphics.PointF;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.navigine.idl.java.AnimationType;
import com.navigine.idl.java.Camera;
import com.navigine.idl.java.InputListener;
import com.navigine.idl.java.MapObjectPickResult;
import com.navigine.idl.java.PickListener;
import com.navigine.idl.java.Point;
import com.navigine.view.LocationView;
import com.navigine.view.TouchInput;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.android.FlutterActivity;
import io.flutter.embedding.android.FlutterFragmentActivity;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;

public class LocationViewController implements
  PlatformView,
  MethodChannel.MethodCallHandler,
  DefaultLifecycleObserver,

  // UserLocationObjectListener,
  // TrafficListener,
  // InputListener,
  // CameraListener,
  // GeoObjectTapListener,
  View.OnLayoutChangeListener
{
  private final LocationView locationView;
  public final Context context;
  public final MethodChannel methodChannel;
  private final NavigineSdkPlugin.LifecycleProvider lifecycleProvider;
  private final CircleMapObjectController circleMapObjectController;
  private final IconMapObjectController iconMapObjectController;
  private final PolylineMapObjectController polylineMapObjectController;

  // private final TrafficLayer trafficLayer;
  // private final UserLocationLayer userLocationLayer;
  // @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  // private PlacemarkMapObjectController userPinController;
  // @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  // private PlacemarkMapObjectController userArrowController;
  // @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  // private CircleMapObjectController userAccuracyCircleController;
  // private final MapObjectCollectionController rootController;
  private boolean disposed = false;
  private MethodChannel.Result initResult;

  @SuppressWarnings({"unchecked", "ConstantConditions", "InflateParams"})
  public LocationViewController(
    int id,
    Context context,
    BinaryMessenger messenger,
    Map<String, Object> params,
    NavigineSdkPlugin.LifecycleProvider lifecycleProvider
  ) {
    this.lifecycleProvider = lifecycleProvider;
    this.context = context;

    if (context instanceof FlutterActivity) {
      locationView = (LocationView) ((FlutterActivity) context).getLayoutInflater().inflate(R.layout.location_view, null);
    } else if (context instanceof FlutterFragmentActivity) {
      locationView = (LocationView) ((FlutterFragmentActivity) context).getLayoutInflater().inflate(R.layout.location_view, null);
    } else {
      locationView = new LocationView(context);
    }

    circleMapObjectController = new CircleMapObjectController(messenger, context, locationView.getLocationWindow());
    iconMapObjectController = new IconMapObjectController(messenger, context, locationView.getLocationWindow());
    polylineMapObjectController = new PolylineMapObjectController(messenger, context, locationView.getLocationWindow());

    methodChannel = new MethodChannel(messenger, "navigine_sdk/navigine_map_" + id);
    methodChannel.setMethodCallHandler(this);

    locationView.addOnLayoutChangeListener(this);

    locationView.getLocationWindow().addInputListener(new InputListener() {
      @Override
      public void onViewTap(PointF pointF) {
        Map<String, Object> arguments = new HashMap<>();
        arguments.put("location", Utils.screenPointToJson(pointF));

        methodChannel.invokeMethod("onTap", arguments);
      }

      @Override
      public void onViewDoubleTap(PointF pointF) {
        Map<String, Object> arguments = new HashMap<>();
        arguments.put("location", Utils.screenPointToJson(pointF));

        methodChannel.invokeMethod("onDoubleTap", arguments);
      }

      @Override
      public void onViewLongTap(PointF pointF) {
        Map<String, Object> arguments = new HashMap<>();
        arguments.put("location", Utils.screenPointToJson(pointF));

        methodChannel.invokeMethod("onLongTap", arguments);
      }
    });

    locationView.getLocationWindow().addPickListener(new PickListener() {
      @Override
      public void onMapObjectPickComplete(MapObjectPickResult mapObjectPickResult, PointF screenPosition) {
        if (mapObjectPickResult == null) {
          return;
        }
        Map<String, Object> arguments = new HashMap<>();

        Map<String, Object> mapObject = new HashMap<>();
        mapObject.put("type", mapObjectPickResult.getMapObject().getType().ordinal());
        mapObject.put("id", mapObjectPickResult.getMapObject().getId());

        Map<String, Object> pickResult = new HashMap<>();
        pickResult.put("locationPoint", Utils.locationPointToJson(mapObjectPickResult.getPoint()));
        pickResult.put("mapObject", mapObject);

        arguments.put("mapObjectPickResult", pickResult);
        arguments.put("screenPosition", Utils.screenPointToJson(screenPosition));

        methodChannel.invokeMethod("onMapObjectPick", arguments);
      }

      @Override
      public void onMapFeaturePickComplete(HashMap<String, String> mapFeaturePickResult, PointF screenPosition) {
        if (mapFeaturePickResult == null) {
          return;
        }

        Map<String, Object> arguments = new HashMap<>();

        arguments.put("mapFeaturePickResult", mapFeaturePickResult);
        arguments.put("screenPosition", Utils.screenPointToJson(screenPosition));

        methodChannel.invokeMethod("onMapFeaturePick", arguments);
      }
    });

    lifecycleProvider.getLifecycle().addObserver(this);
  }

  @Override
  public View getView() {
    return locationView;
  }

  @Override
  public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "waitForInit":
        if (locationView.getWidth() == 0 || locationView.getHeight() == 0) {
          initResult = result;
        } else {
          result.success(null);
        }
        break;
      case "setSublocationId":
        setSublocationId(call);
        result.success(null);
        break;
      case "removeAllMapObjects":
        removeAllMapObjects();
        result.success(null);
        break;
      case "addCircleMapObject":
        result.success(addCircleMapObject(call));
        break;
      case "removeCircleMapObject":
        result.success(removeCircleMapObject(call));
        break;
      case "addIconMapObject":
        result.success(addIconMapObject(call));
        break;
      case "removeIconMapObject":
        result.success(removeIconMapObject(call));
        break;
      case "addPolylineMapObject":
        result.success(addPolylineMapObject(call));
        break;
      case "removePolylineMapObject":
        result.success(removePolylineMapObject(call));
        break;
      case "screenPositionToMeters":
        result.success(screenPositionToMeters(call));
        break;
      case "metersToScreenPosition":
        result.success(metersToScreenPosition(call));
        break;
      case "pickMapObjectAt":
        pickMapObjectAt(call);
        result.success(null);
        break;
      case "pickMapFeatureAt":
        pickMapFeatureAt(call);
        result.success(null);
        break;
      case "applyFilter":
        applyFilter(call);
        result.success(null);
        break;
      case "setMinZoomFactor":
        setMinZoomFactor(call);
        result.success(null);
        break;
      case "getMinZoomFactor":
        result.success(getMinZoomFactor(call));
        break;
      case "setMaxZoomFactor":
        setMaxZoomFactor(call);
        result.success(null);
        break;
      case "getMaxZoomFactor":
        result.success(getMaxZoomFactor(call));
        break;
      case "setZoomFactor":
        setZoomFactor(call);
        result.success(null);
        break;
      case "getZoomFactor":
        result.success(getZoomFactor(call));
        break;
      case "setCamera":
        setCamera(call);
        result.success(null);
        break;
      case "getCamera":
        result.success(getCamera(call));
        break;
      case "flyToCamera":
        flyToCamera(call);
        result.success(null);
        break;

      default:
        result.notImplemented();
        break;
    }
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setSublocationId(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    locationView.getLocationWindow().setSublocationId((Integer) params.get("sublocationId"));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void removeAllMapObjects() {
    locationView.getLocationWindow().removeAllMapObjects();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public int addCircleMapObject(MethodCall call) {
    return this.circleMapObjectController.addCircleMapObject();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean removeCircleMapObject(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    return this.circleMapObjectController.removeCircleMapObject((Integer) params.get("id"));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public int addIconMapObject(MethodCall call) {
    return this.iconMapObjectController.addIconMapObject();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean removeIconMapObject(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    return this.iconMapObjectController.removeIconMapObject((Integer) params.get("id"));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public int addPolylineMapObject(MethodCall call) {
    return this.polylineMapObjectController.addPolylineMapObject();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean removePolylineMapObject(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    return this.polylineMapObjectController.removePolylineMapObject((Integer) params.get("id"));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public Map<String, Object> screenPositionToMeters(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;

    PointF screenPoint = Utils.screenPointFromJson((Map<String, Object>) params.get("point"));
    return Utils.pointToJson(locationView.getLocationWindow().screenPositionToMeters(screenPoint));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public Map<String, Object> metersToScreenPosition(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;

    Point point = Utils.pointFromJson((Map<String, Object>) params.get("point"));

    Boolean clip = (Boolean)params.get("clip");
    return Utils.screenPointToJson(locationView.getLocationWindow().metersToScreenPosition(point, clip));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void pickMapObjectAt(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;

    PointF point = Utils.screenPointFromJson((Map<String, Object>) params.get("point"));

    locationView.getLocationWindow().pickMapObjectAt(point);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void pickMapFeatureAt(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;

    PointF point = Utils.screenPointFromJson((Map<String, Object>) params.get("point"));

    locationView.getLocationWindow().pickMapFeatureAt(point);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void applyFilter(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    locationView.getLocationWindow().applyFilter(
            (String) params.get("filter"),
            (String) params.get("layer"));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setMinZoomFactor(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    Double minZoomFactor = (Double) params.get("minZoomFactor");
    locationView.getLocationWindow().setMinZoomFactor(minZoomFactor.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public float getMinZoomFactor(MethodCall call) {
    return locationView.getLocationWindow().getMinZoomFactor();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setMaxZoomFactor(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    Double maxZoomFactor = (Double) params.get("maxZoomFactor");
    locationView.getLocationWindow().setMaxZoomFactor(maxZoomFactor.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public float getMaxZoomFactor(MethodCall call) {
    return locationView.getLocationWindow().getMaxZoomFactor();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setZoomFactor(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    Double zoomFactor = (Double) params.get("zoomFactor");
    locationView.getLocationWindow().setZoomFactor(zoomFactor.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public float getZoomFactor(MethodCall call) {
    return locationView.getLocationWindow().getZoomFactor();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setCamera(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    locationView.getLocationWindow().setCamera(Utils.cameraFromJson((Map<String, Object>) params.get("camera")));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public Map<String, Object> getCamera(MethodCall call) {
    return Utils.cameraToJson(locationView.getLocationWindow().getCamera());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void flyToCamera(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    Camera camera = Utils.cameraFromJson((Map<String, Object>) params.get("camera"));
    Integer duration = (Integer) params.get("duration");
    locationView.getLocationWindow().flyTo(camera, duration, null);
  }

  @Override
  public void dispose() {
    if (disposed) {
      return;
    }

    disposed = true;
    methodChannel.setMethodCallHandler(null);

    Lifecycle lifecycle = lifecycleProvider.getLifecycle();
    if (lifecycle != null) {
      lifecycle.removeObserver(this);
    }
  }

  @Override
  public void onCreate(@NonNull LifecycleOwner owner) {}

  @Override
  public void onStart(@NonNull LifecycleOwner owner) {
    if (disposed) {
      return;
    }
  }

  @Override
  public void onResume(@NonNull LifecycleOwner owner) {}

  @Override
  public void onPause(@NonNull LifecycleOwner owner) {}

  @Override
  public void onStop(@NonNull LifecycleOwner owner) {
    if (disposed) {
      return;
    }
  }

  @Override
  public void onDestroy(@NonNull LifecycleOwner owner) {
    owner.getLifecycle().removeObserver(this);
  }

  @Override
  public void onLayoutChange(View view, int i, int i1, int i2, int i3, int i4, int i5, int i6, int i7) {
    if (initResult != null) {
      initResult.success(null);
      initResult = null;
    }
  }
}

File: ios/Classes/LocationViewController.swift
import CoreLocation
import Flutter
import UIKit
import Navigine

public class LocationViewController:
    NSObject,
    FlutterPlatformView,
    NCPickListener,
    NCInputListener
{
    public let methodChannel: FlutterMethodChannel!
    public let pluginRegistrar: FlutterPluginRegistrar!

    private let locationView: FLNCLocationView
    private let circleMapObjectController: CircleMapObjectController
    private let iconMapObjectController: IconMapObjectController
    private let polylineMapObjectController: PolylineMapObjectController

    public required init(id: Int64, frame: CGRect, registrar: FlutterPluginRegistrar, params: [String: Any]) {
        self.pluginRegistrar = registrar
        self.locationView = FLNCLocationView(frame: frame)
        self.methodChannel = FlutterMethodChannel(
            name: "navigine_sdk/navigine_map_\(id)",
            binaryMessenger: registrar.messenger()
        )
        self.circleMapObjectController = CircleMapObjectController(registrar: pluginRegistrar, locationView: locationView)
        self.iconMapObjectController = IconMapObjectController(registrar: pluginRegistrar, locationView: locationView)
        self.polylineMapObjectController = PolylineMapObjectController(registrar: pluginRegistrar, locationView: locationView)

        super.init()

        weak var weakSelf = self
        self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })

        self.locationView.locationWindow.add(self as NCPickListener)
        self.locationView.locationWindow.add(self as NCInputListener)
    }

    public func view() -> UIView {
        return self.locationView
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "waitForInit":
            if (locationView.frame.isEmpty) {
                locationView.initResult = result
            } else {
                result(nil)
            }
        case "setSublocationId":
            setSublocationId(call)
            result(nil)
        case "removeAllMapObjects":
            removeAllMapObjects(call)
            result(nil)
        case "addCircleMapObject":
            result(addCircleMapObject(call))
        case "removeCircleMapObject":
            result(removeCircleMapObject(call))
        case "addIconMapObject":
            result(addIconMapObject(call))
        case "removeIconMapObject":
            result(removeIconMapObject(call))
        case "addPolylineMapObject":
            result(addPolylineMapObject(call))
        case "removePolylineMapObject":
            result(removePolylineMapObject(call))
        case "screenPositionToMeters":
            result(screenPositionToMeters(call))
        case "metersToScreenPosition":
            result(metersToScreenPosition(call))
        case "pickMapObjectAt":
            pickMapObjectAt(call)
            result(nil)
        case "pickMapFeatureAt":
            pickMapFeatureAt(call)
            result(nil)
        case "applyFilter":
            applyFilter(call)
            result(nil)
        case "setMinZoomFactor":
            setMinZoomFactor(call)
            result(nil)
        case "getMinZoomFactor":
            result(getMinZoomFactor(call))
        case "setMaxZoomFactor":
            setMaxZoomFactor(call)
            result(nil)
        case "getMaxZoomFactor":
            result(getMaxZoomFactor(call))
        case "setZoomFactor":
            setZoomFactor(call)
            result(nil)
        case "getZoomFactor":
            result(getZoomFactor(call))
        case "setCamera":
            setCamera(call)
            result(nil)
        case "getCamera":
            result(getCamera(call))
        case "flyToCamera":
            flyToCamera(call)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func setSublocationId(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]

        locationView.locationWindow.setSublocationId(params["sublocationId"] as! Int32)
    }

    public func removeAllMapObjects(_ call: FlutterMethodCall) {
        locationView.locationWindow.removeAllMapObjects()
    }

    public func addCircleMapObject(_ call: FlutterMethodCall) -> Int32? {
        return self.circleMapObjectController.addCircleMapObject()
    }

    public func removeCircleMapObject(_ call: FlutterMethodCall) -> Bool? {
        let params = call.arguments as! [String: Any]
        return self.circleMapObjectController.removeCircleMapObject(id: params["id"] as! Int32)
    }

    public func addIconMapObject(_ call: FlutterMethodCall) -> Int32? {
        return self.iconMapObjectController.addIconMapObject()
    }

    public func removeIconMapObject(_ call: FlutterMethodCall) -> Bool? {
        let params = call.arguments as! [String: Any]
        return self.iconMapObjectController.removeIconMapObject(id: params["id"] as! Int32)
    }

    public func addPolylineMapObject(_ call: FlutterMethodCall) -> Int32? {
        return self.polylineMapObjectController.addPolylineMapObject()
    }

    public func removePolylineMapObject(_ call: FlutterMethodCall) -> Bool? {
        let params = call.arguments as! [String: Any]
        return self.polylineMapObjectController.removePolylineMapObject(id: params["id"] as! Int32)
    }

    public func screenPositionToMeters(_ call: FlutterMethodCall) -> [String: Any?] {
        let params = call.arguments as! [String: Any]
        let screenPoint = Utils.screenPointFromJson(params["point"] as! [String: NSNumber])
        return Utils.pointToJson(locationView.locationWindow.screenPosition(toMeters: screenPoint))
    }

    public func metersToScreenPosition(_ call: FlutterMethodCall) -> [String: Any?] {
        let params = call.arguments as! [String: Any]

        let point = Utils.pointFromJson(params["point"] as! [String: NSNumber])
        let clip = params["clip"] as! Bool
        return Utils.screenPointToJson(locationView.locationWindow.meters(toScreenPosition: point, clipToViewport: clip))
    }

    public func pickMapObjectAt(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        locationView.locationWindow.pickMapObject(at: Utils.screenPointFromJson(params["point"] as! [String: NSNumber]))
    }

    public func pickMapFeatureAt(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        locationView.locationWindow.pickMapFeature(at: Utils.screenPointFromJson(params["point"] as! [String: NSNumber]))
    }

    public func applyFilter(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        locationView.locationWindow.applyFilter(params["filter"] as! String, layer: params["layer"] as! String)
    }

    public func setMinZoomFactor(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let minZoomFactor = params["minZoomFactor"] as! Float
        locationView.locationWindow.minZoomFactor = minZoomFactor
    }

    public func getMinZoomFactor(_ call: FlutterMethodCall) -> Float {
        return Float(locationView.locationWindow.minZoomFactor)
    }

    public func setMaxZoomFactor(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let maxZoomFactor = params["maxZoomFactor"] as! Float
        locationView.locationWindow.maxZoomFactor = maxZoomFactor
    }

    public func getMaxZoomFactor(_ call: FlutterMethodCall) -> Float {
        return Float(locationView.locationWindow.maxZoomFactor)
    }

    public func setZoomFactor(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let zoomFactor = params["zoomFactor"] as! Float
        locationView.locationWindow.zoomFactor = zoomFactor
    }

    public func getZoomFactor(_ call: FlutterMethodCall) -> Float {
        return Float(locationView.locationWindow.zoomFactor)
    }

    public func setCamera(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        locationView.locationWindow.camera = Utils.cameraFromJson(params["camera"] as! [String: Any])
    }

    public func getCamera(_ call: FlutterMethodCall) -> [String: Any?] {
        return Utils.cameraToJson(locationView.locationWindow.camera)
    }

    public func flyToCamera(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let camera = Utils.cameraFromJson(params["camera"] as! [String: Any])
        let duration = params["duration"] as! NSNumber
        locationView.locationWindow.fly(to: camera, duration: duration.int32Value) { finished in
            let arguments: [String: Bool] = [
                "finished": finished
            ]
            self.methodChannel.invokeMethod("onCameraAnimation", arguments: arguments)
        }
    }

    private func hasLocationPermission() -> Bool {
        if CLLocationManager.locationServicesEnabled() {
            switch CLLocationManager.authorizationStatus() {
            case .notDetermined, .restricted, .denied:
                return false
            case .authorizedAlways, .authorizedWhenInUse:
                return true
            default:
                return false
            }
        } else {
            return false
        }
    }

    public func onMapObjectPickComplete(_ mapObjectPickResult: NCMapObjectPickResult?, screenPosition: CGPoint) {
        if mapObjectPickResult == nil {
            return
        }

        let arguments: [String: Any?] = [
            "mapObjectPickResult": [
                "locationPoint": Utils.locationPointToJson(mapObjectPickResult!.point),
                "mapObject": [
                    "id": mapObjectPickResult!.mapObject?.getId() as Any,
                    "type": mapObjectPickResult!.mapObject?.getType().rawValue
                ]
            ],
            "screenPosition" : Utils.screenPointToJson(screenPosition)
        ]

        methodChannel.invokeMethod("onMapObjectPick", arguments: arguments)
    }

    public func onMapFeaturePickComplete(_ mapFeaturePickResult: [String : String], screenPosition: CGPoint) {
        let arguments: [String: Any?] = [
            "mapFeaturePickResult": mapFeaturePickResult,
            "screenPosition" : Utils.screenPointToJson(screenPosition)
        ]

        methodChannel.invokeMethod("onMapFeaturePick", arguments: arguments)
    }

    public func onViewTap(_ screenPoint: CGPoint) {
        let arguments: [String: Any?] = [
            "location": Utils.screenPointToJson(screenPoint)
        ]
        methodChannel.invokeMethod("onTap", arguments: arguments)
    }

    public func onViewDoubleTap(_ screenPoint: CGPoint) {
        let arguments: [String: Any?] = [
            "location": Utils.screenPointToJson(screenPoint)
        ]
        methodChannel.invokeMethod("onDoubleTap", arguments: arguments)
    }

    public func onViewLongTap(_ screenPoint: CGPoint) {
        let arguments: [String: Any?] = [
            "location": Utils.screenPointToJson(screenPoint)
        ]
        methodChannel.invokeMethod("onLongTap", arguments: arguments)
    }

    // Fix https://github.com/flutter/flutter/issues/67514
    internal class FLNCLocationView: NCLocationView {
        public var initResult: FlutterResult?

        override var frame: CGRect {
            didSet {
                if initResult != nil {
                    initResult!(nil)
                    initResult = nil
                }
            }
        }
    }
}
